zcash_client_backend/src/welding_rig.rs

```diff
@@ -265,17 +201,43 @@ pub fn scan_block<P: consensus::Parameters, K: ScanningKey>(
                 })
                 .collect();
 
-            for (index, output) in tx.outputs.into_iter().enumerate() {
+            let decoded = &tx
+                .outputs
+                .into_iter()
+                .map(|output| {
+                    (
+                        SaplingDomain::for_height(params.clone(), block_height),
+                        CompactOutputDescription::try_from(output)
+                            .expect("Invalid output found in compact block decoding."),
+                    )
+                })
+                .collect::<Vec<_>>();
+
+            let vks = vks
+                .iter()
+                .flat_map(|(a, k)| {
+                    k.to_sapling_keys()
+                        .into_iter()
+                        .map(move |(ivk, nk)| (**a, ivk, nk))
+                })
+                .collect::<Vec<_>>();
+
+            let ivks = vks
+                .iter()
+                .map(|(_, ivk, _)| (*ivk).clone())
+                .collect::<Vec<_>>();
+
+            let decrypted = batch::try_compact_note_decryption(&ivks, decoded);
+
+            for (index, ((_, output), dec_output)) in decoded.iter().zip(decrypted).enumerate() {
                 // Grab mutable references to new witnesses from previous outputs
                 // in this transaction so that we can update them. Scoped so we
                 // don't hold mutable references to shielded_outputs for too long.
                 let new_witnesses: Vec<_> = shielded_outputs
                     .iter_mut()
-                    .map(|output| &mut output.witness)
+                    .map(|out| &mut out.witness)
```

Comment:
Nit: this line didn't need to be changed.

Reply:
I didn't like the shadowing, it confused me.